File: js/hqc/src/gf2.js
/**
 * GF(2) polynomial arithmetic using Uint32Array word packing.
 *
 * Polynomials over GF(2) are packed into Uint32Array words (32 bits each).
 * Arithmetic is in GF(2)[x] / (x^n - 1).
 */

const WORD_BITS = 32;

/**
 * Allocate a zero vector with enough words for nBits.
 * @param {number} nBits
 * @returns {Uint32Array}
 */
export function vectAlloc(nBits) {
  return new Uint32Array(Math.ceil(nBits / WORD_BITS));
}

/**
 * Compute out = a XOR b (polynomial addition in GF(2)).
 * @param {Uint32Array} a
 * @param {Uint32Array} b
 * @returns {Uint32Array}
 */
export function vectAdd(a, b) {
  const n = Math.max(a.length, b.length);
  const out = new Uint32Array(n);
  for (let i = 0; i < a.length; i++) out[i] = a[i];
  for (let i = 0; i < b.length; i++) out[i] ^= b[i];
  return out;
}

/**
 * Set bit at position pos in vector v.
 * @param {Uint32Array} v
 * @param {number} pos
 */
export function vectSetBit(v, pos) {
  v[pos >>> 5] |= 1 << (pos & 31);
}

/**
 * Get bit at position pos in vector v.
 * @param {Uint32Array} v
 * @param {number} pos
 * @returns {number} 0 or 1
 */
export function vectGetBit(v, pos) {
  return (v[pos >>> 5] >>> (pos & 31)) & 1;
}

/**
 * Hamming weight of a GF(2) vector.
 * @param {Uint32Array} v
 * @returns {number}
 */
export function vectWeight(v) {
  let w = 0;
  for (let i = 0; i < v.length; i++) {
    let x = v[i];
    // Kernighan's bit counting
    while (x) {
      x &= x - 1;
      w++;
    }
  }
  return w;
}

/**
 * Convert Uint32Array to bytes (little-endian).
 * @param {Uint32Array} v
 * @param {number} nBytes
 * @returns {Uint8Array}
 */
export function vectToBytes(v, nBytes) {
  const out = new Uint8Array(nBytes);
  for (let i = 0; i < v.length && i * 4 < nBytes; i++) {
    const remaining = nBytes - i * 4;
    const word = v[i];
    if (remaining >= 4) {
      out[i * 4] = word & 0xFF;
      out[i * 4 + 1] = (word >>> 8) & 0xFF;
      out[i * 4 + 2] = (word >>> 16) & 0xFF;
      out[i * 4 + 3] = (word >>> 24) & 0xFF;
    } else {
      for (let b = 0; b < remaining; b++) {
        out[i * 4 + b] = (word >>> (b * 8)) & 0xFF;
      }
    }
  }
  return out;
}

/**
 * Convert bytes to Uint32Array (little-endian).
 * @param {Uint8Array} data
 * @param {number} nWords
 * @returns {Uint32Array}
 */
export function vectFromBytes(data, nWords) {
  const v = new Uint32Array(nWords);
  for (let i = 0; i < nWords; i++) {
    const start = i * 4;
    if (start >= data.length) break;
    let word = 0;
    const end = Math.min(start + 4, data.length);
    for (let b = start; b < end; b++) {
      word |= data[b] << ((b - start) * 8);
    }
    v[i] = word >>> 0;
  }
  return v;
}

/**
 * Resize vector to exactly nBits, masking the last word.
 * @param {Uint32Array} v
 * @param {number} nBits
 * @returns {Uint32Array}
 */
export function vectResize(v, nBits) {
  const nWords = Math.ceil(nBits / WORD_BITS);
  const out = new Uint32Array(nWords);
  const copyLen = Math.min(v.length, nWords);
  for (let i = 0; i < copyLen; i++) out[i] = v[i];
  const rem = nBits & 31;
  if (rem !== 0 && nWords > 0) {
    out[nWords - 1] &= (1 << rem) - 1;
  }
  return out;
}

/**
 * Constant-time equality check. Returns 1 if equal, 0 otherwise.
 * @param {Uint32Array} a
 * @param {Uint32Array} b
 * @returns {number}
 */
export function vectEqual(a, b) {
  let diff = 0;
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) diff |= a[i] ^ b[i];
  for (let i = n; i < a.length; i++) diff |= a[i];
  for (let i = n; i < b.length; i++) diff |= b[i];
  // Collapse diff to a single bit
  diff |= diff >>> 16;
  diff |= diff >>> 8;
  diff |= diff >>> 4;
  diff |= diff >>> 2;
  diff |= diff >>> 1;
  return 1 - (diff & 1);
}

/**
 * Carryless multiplication of two 32-bit words.
 * Returns [lo, hi] such that a * b = hi<<32 | lo in GF(2).
 * @param {number} a
 * @param {number} b
 * @returns {[number, number]}
 */
function baseMul(a, b) {
  // Use BigInt for the carryless multiply to avoid precision issues
  let lo = 0n;
  const ab = BigInt(a >>> 0);
  const bb = BigInt(b >>> 0);

  for (let i = 0; i < 32; i++) {
    if ((ab >> BigInt(i)) & 1n) {
      lo ^= bb << BigInt(i);
    }
  }

  return [Number(lo & 0xFFFFFFFFn), Number((lo >> 32n) & 0xFFFFFFFFn)];
}

/**
 * Schoolbook polynomial multiplication of two GF(2) polynomials.
 * @param {Uint32Array} a - size sizeA words
 * @param {number} sizeA
 * @param {Uint32Array} b - size sizeB words
 * @param {number} sizeB
 * @returns {Uint32Array} - size sizeA+sizeB words
 */
function schoolbookMul(a, sizeA, b, sizeB) {
  const out = new Uint32Array(sizeA + sizeB);
  for (let i = 0; i < sizeA; i++) {
    if (a[i] === 0) continue;
    for (let j = 0; j < sizeB; j++) {
      if (b[j] === 0) continue;
      const [lo, hi] = baseMul(a[i], b[j]);
      out[i + j] ^= lo;
      out[i + j + 1] ^= hi;
    }
  }
  return out;
}

/**
 * Compute out = a * b mod (x^n - 1) in GF(2)[x].
 * @param {Uint32Array} a
 * @param {Uint32Array} b
 * @param {number} n
 * @returns {Uint32Array}
 */
export function vectMul(a, b, n) {
  const nWords = Math.ceil(n / WORD_BITS);

  // Pad and mask inputs
  const aPad = new Uint32Array(nWords);
  const bPad = new Uint32Array(nWords);
  const copyA = Math.min(a.length, nWords);
  const copyB = Math.min(b.length, nWords);
  for (let i = 0; i < copyA; i++) aPad[i] = a[i];
  for (let i = 0; i < copyB; i++) bPad[i] = b[i];

  const rem = n & 31;
  if (rem !== 0) {
    aPad[nWords - 1] &= (1 << rem) - 1;
    bPad[nWords - 1] &= (1 << rem) - 1;
  }

  // Full product
  const prod = schoolbookMul(aPad, nWords, bPad, nWords);

  // Reduce mod (x^n - 1)
  const out = new Uint32Array(nWords);
  for (let i = 0; i < nWords; i++) out[i] = prod[i];

  const wordOff = (n / WORD_BITS) | 0;

  if (rem === 0) {
    for (let i = 0; i < nWords; i++) {
      if (wordOff + i < 2 * nWords) {
        out[i] ^= prod[wordOff + i];
      }
    }
  } else {
    for (let i = 0; i < nWords; i++) {
      const idx = wordOff + i;
      if (idx < 2 * nWords) {
        out[i] ^= prod[idx] >>> rem;
      }
      if (idx + 1 < 2 * nWords) {
        out[i] ^= prod[idx + 1] << (WORD_BITS - rem);
      }
    }
  }

  // Mask last word
  if (rem !== 0) {
    out[nWords - 1] &= (1 << rem) - 1;
  }

  return out;
}

/**
 * Extract nBits bits from src starting at bitOffset.
 * @param {Uint32Array} src
 * @param {number} bitOffset
 * @param {number} nBits
 * @returns {Uint32Array}
 */
export function extractBits(src, bitOffset, nBits) {
  const nWords = Math.ceil(nBits / WORD_BITS);
  const out = new Uint32Array(nWords);

  const srcWord = (bitOffset / WORD_BITS) | 0;
  const srcBit = bitOffset & 31;

  if (srcBit === 0) {
    for (let i = 0; i < nWords && srcWord + i < src.length; i++) {
      out[i] = src[srcWord + i];
    }
  } else {
    for (let i = 0; i < nWords; i++) {
      const idx = srcWord + i;
      if (idx < src.length) {
        out[i] = src[idx] >>> srcBit;
      }
      if (idx + 1 < src.length) {
        out[i] |= src[idx + 1] << (WORD_BITS - srcBit);
      }
    }
  }

  // Mask last word
  const remBits = nBits & 31;
  if (remBits !== 0 && nWords > 0) {
    out[nWords - 1] &= (1 << remBits) - 1;
  }

  return out;
}

File: rust/hqc/src/gf2.rs
/// GF(2) polynomial arithmetic: polynomials over GF(2) packed into `Vec<u64>` words.
///
/// Each polynomial has at most n bits. Arithmetic is in GF(2)[x]/(x^n - 1).

/// Polynomial addition in GF(2): out = a XOR b.
pub fn vect_add(a: &[u64], b: &[u64]) -> Vec<u64> {
    let n = a.len().max(b.len());
    let mut out = vec![0u64; n];
    for i in 0..a.len() {
        out[i] = a[i];
    }
    for i in 0..b.len() {
        out[i] ^= b[i];
    }
    out
}

/// Set bit at position `pos` in the vector `v`.
pub fn vect_set_bit(v: &mut [u64], pos: usize) {
    v[pos / 64] |= 1u64 << (pos % 64);
}

/// Get bit at position `pos` in the vector `v`.
pub fn vect_get_bit(v: &[u64], pos: usize) -> u64 {
    (v[pos / 64] >> (pos % 64)) & 1
}

/// Returns the Hamming weight of a GF(2) vector.
pub fn vect_weight(v: &[u64]) -> usize {
    v.iter().map(|w| w.count_ones() as usize).sum()
}

/// Converts a u64 vector to bytes (little-endian).
pub fn vect_to_bytes(v: &[u64], n_bytes: usize) -> Vec<u8> {
    let mut out = vec![0u8; n_bytes];
    for i in 0..v.len() {
        let start = i * 8;
        if start >= n_bytes {
            break;
        }
        let bytes = v[i].to_le_bytes();
        let remaining = n_bytes - start;
        let copy_len = remaining.min(8);
        out[start..start + copy_len].copy_from_slice(&bytes[..copy_len]);
    }
    out
}

/// Converts bytes to a u64 vector (little-endian).
pub fn vect_from_bytes(data: &[u8], n_words: usize) -> Vec<u64> {
    let mut v = vec![0u64; n_words];
    for i in 0..n_words {
        let start = i * 8;
        if start >= data.len() {
            break;
        }
        let end = (start + 8).min(data.len());
        let mut buf = [0u8; 8];
        buf[..end - start].copy_from_slice(&data[start..end]);
        v[i] = u64::from_le_bytes(buf);
    }
    v
}

/// Returns a copy of v truncated/masked to exactly `n_bits` bits.
pub fn vect_resize(v: &[u64], n_bits: usize) -> Vec<u64> {
    let n_words = (n_bits + 63) / 64;
    let mut out = vec![0u64; n_words];
    let copy_len = out.len().min(v.len());
    out[..copy_len].copy_from_slice(&v[..copy_len]);
    let rem = n_bits % 64;
    if rem != 0 && n_words > 0 {
        out[n_words - 1] &= (1u64 << rem) - 1;
    }
    out
}

/// Constant-time equality: returns 1 if a == b, 0 otherwise.
pub fn vect_equal(a: &[u64], b: &[u64]) -> usize {
    let mut diff = 0u64;
    let n = a.len().min(b.len());
    for i in 0..n {
        diff |= a[i] ^ b[i];
    }
    for i in n..a.len() {
        diff |= a[i];
    }
    for i in n..b.len() {
        diff |= b[i];
    }
    let mut d = diff | (diff >> 32);
    d |= d >> 16;
    d |= d >> 8;
    d |= d >> 4;
    d |= d >> 2;
    d |= d >> 1;
    1 - (d & 1) as usize
}

/// Carryless multiplication of two 64-bit words.
/// Returns (lo, hi) such that a * b = hi<<64 | lo in GF(2).
fn base_mul(a: u64, b: u64) -> (u64, u64) {
    let mut lo = 0u64;
    let mut hi = 0u64;

    for i in 0..64 {
        if (a >> i) & 1 == 0 {
            continue;
        }
        if i == 0 {
            lo ^= b;
        } else {
            lo ^= b << i;
            hi ^= b >> (64 - i);
        }
    }

    (lo, hi)
}

/// Schoolbook polynomial multiplication of two GF(2) polynomials.
fn schoolbook_mul(a: &[u64], size_a: usize, b: &[u64], size_b: usize) -> Vec<u64> {
    let mut out = vec![0u64; size_a + size_b];
    for i in 0..size_a {
        if a[i] == 0 {
            continue;
        }
        for j in 0..size_b {
            if b[j] == 0 {
                continue;
            }
            let (lo, hi) = base_mul(a[i], b[j]);
            out[i + j] ^= lo;
            out[i + j + 1] ^= hi;
        }
    }
    out
}

/// Computes out = a * b mod (x^n - 1) in GF(2)[x].
pub fn vect_mul(a: &[u64], b: &[u64], n: usize) -> Vec<u64> {
    let n_words = (n + 63) / 64;

    // Pad and mask inputs
    let mut a_pad = vec![0u64; n_words];
    let mut b_pad = vec![0u64; n_words];
    let copy_a = a.len().min(n_words);
    let copy_b = b.len().min(n_words);
    a_pad[..copy_a].copy_from_slice(&a[..copy_a]);
    b_pad[..copy_b].copy_from_slice(&b[..copy_b]);

    let rem = n % 64;
    if rem != 0 {
        a_pad[n_words - 1] &= (1u64 << rem) - 1;
        b_pad[n_words - 1] &= (1u64 << rem) - 1;
    }

    // Full product
    let prod = schoolbook_mul(&a_pad, n_words, &b_pad, n_words);

    // Reduce mod (x^n - 1)
    let mut out = vec![0u64; n_words];
    out[..n_words].copy_from_slice(&prod[..n_words]);

    let word_off = n / 64;

    if rem == 0 {
        for i in 0..n_words {
            if word_off + i < 2 * n_words {
                out[i] ^= prod[word_off + i];
            }
        }
    } else {
        for i in 0..n_words {
            let idx = word_off + i;
            if idx < 2 * n_words {
                out[i] ^= prod[idx] >> rem;
            }
            if idx + 1 < 2 * n_words {
                out[i] ^= prod[idx + 1] << (64 - rem);
            }
        }
    }

    // Mask last word
    if rem != 0 {
        out[n_words - 1] &= (1u64 << rem) - 1;
    }

    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_vect_add() {
        let a = vec![0xAAAAu64, 0x5555];
        let b = vec![0x5555u64, 0xAAAA];
        let c = vect_add(&a, &b);
        assert_eq!(c[0], 0xFFFF);
        assert_eq!(c[1], 0xFFFF);
    }

    #[test]
    fn test_vect_mul_identity() {
        let one = vec![1u64, 0];
        let d = vec![0xDEADBEEFCAFEBABEu64, 0x1234567890ABCDEF];
        let r = vect_mul(&d, &one, 128);
        assert_eq!(r[0], d[0]);
        assert_eq!(r[1], d[1]);
    }

    #[test]
    fn test_vect_bytes_roundtrip() {
        let v = vec![0x0123456789ABCDEFu64, 0xFEDCBA9876543210];
        let bytes = vect_to_bytes(&v, 16);
        let v2 = vect_from_bytes(&bytes, 2);
        assert_eq!(v, v2);
    }
}
